pkg/cmd/roachtest/tests/mixed_version_sql_stats.go

```diff
@@ -203,7 +203,9 @@ func (s *sqlStatsRequestHelper) requestSQLStats(
 	for _, addr := range adminUIAddrs {
 		url := getCombinedStatementStatsURL(addr, statsType, requestedRange)
 		statsResponse := &serverpb.StatementsResponse{}
-		if err := s.client.GetJSON(ctx, url, statsResponse, httputil.IgnoreUnknownFields()); err != nil {
+		if err := retry.ForDuration(10*time.Second, func() error {
```

Comment:
Why do we need the retry logic? What kinds of spurious failures do we expect here?

Reply:
Mostly motivated from seeing it applied here: https://github.com/cockroachdb/cockroach/blob/7c3d18ae4a3024ddf48a87e96098672334e1bebc/pkg/cmd/roachtest/tests/status_server.go#L50, where it seems like the status server was not yet ready to serve requests.

Reply:
Hrm, but that change was done in 2018 (#29492), does it still apply today? Have we seen this test flake because the status server was not ready to handle requests?

Reply:
Okay that seems pretty outdated. So far this test has failed twice in the nightly run at this step, but with different error messages:

1. `Get "https://3.147.89.154:26258/_status/combinedstmts?fetch_mode.stats_type=0&start=0&end=100&limit=10": EOF` [[ref](https://github.com/cockroachdb/cockroach/issues/123278#issue-2270771970)]

2. `roachtestutil.addCookie: unable to extract sessionID: roachtestutil.GetSessionID: failed to authenticate: COMMAND_PROBLEM: exit status 1` [[ref](https://github.com/cockroachdb/cockroach/issues/123278#issuecomment-2088532991)]

Do these point to an issue with the mixed version test or an issue within roachtest? Is there guidance on restricting the test to certain cloud providers? Running the test locally passed so these failures seemed sporadic to me.

Reply:
That second issue seems the same as https://github.com/cockroachdb/cockroach/issues/119064. You weren't running into it before because you had `mixedversion.MaxUpgrades(2)` but that was removed.

I think we need to either add that back in or run it in insecure mode, neither of which are great options. The former wouldn't solve the issue if we backport this test, and the latter would mean we have 0 mixed version roachtests that run on secure mode. 

FWIW I only noticed https://github.com/cockroachdb/cockroach/pull/122468 because this test was running on secure mode so I'm not convinced disabling it is the easy answer. But maybe with 23.1 support dropping soon it doesn't matter 🤷

Reply:
For failure 2, I see: `ERROR: column "user_id" does not exist` when trying to auth on `23.1`. Since I removed the version gate for that, I think the fix is for the mvt framework to also drop support for 23.1 on master?

Reply:
> I think the fix is for the mvt framework to also drop support for 23.1 on master?

Ah my bad, I didn't realize this is why the PR had `MaxUpgrades(2)` originally. I forgot the details on when the auth stuff works. I suggest passing the `MinimumSupportedVersion("v23.1.0")` option to the test, which will make sure user functions will only run once the cluster is at least that version.

> Do these point to an issue with the mixed version test or an issue within roachtest?

I think neither in this case? Failure 2 is a limitation in cockroach with authenticating in mixed-version state, as I understand it. It can be worked around by specifying the minimum version as I mentioned above.

Issue 1 actually seems to be a bug in the stats code -- n4 crashed with this error (only a few frames here for clarity):

```
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506  a panic has occurred!
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +runtime error: invalid memory address or nil pointer dereference
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +(1) attached stack trace
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  -- stack trace:
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | runtime.gopanic
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | 	GOROOT/src/runtime/panic.go:884
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | runtime.panicmem
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | 	GOROOT/src/runtime/panic.go:260
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | runtime.sigpanic
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | 	GOROOT/src/runtime/signal_unix.go:835
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | github.com/cockroachdb/cockroach/pkg/server.activityTablesHaveFullData
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | 	github.com/cockroachdb/cockroach/pkg/server/combined_statement_stats.go:232
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | github.com/cockroachdb/cockroach/pkg/server.getCombinedStatementStats
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | 	github.com/cockroachdb/cockroach/pkg/server/combined_statement_stats.go:100
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | github.com/cockroachdb/cockroach/pkg/server.(*statusServer).CombinedStatementStats
E240430 07:54:10.992587 21115 1@util/log/logcrash/crash_reporting.go:188 ⋮ [-] 506 +  | 	github.com/cockroachdb/cockroach/pkg/server/combined_statement_stats.go:65
```

Reply:
Granted, the error message could be more explicit (and it could be if the monitor had noticed the process death soon enough, but the client saw the EOF first, and that was the error reported). A client seeing an EOF from a connection often means that the server crashed, FWIW.

> Running the test locally passed so these failures seemed sporadic to me.

As a bit of context, there's a lot of randomization happening in these tests so passing locally doesn't mean that test failures are flakes or spurious necessarily.